modules/internal/src/main/scala/higherkindness/mu/rpc/internal/serviceImpl.scala

```diff
@@ -338,47 +386,229 @@ object serviceImpl {
+
+          case (_: UnaryTpe, _: MonixObservableTpe) =>
+            q"""for {
+                request   <- msg.as[${operation.request.safeInner}]
+                responses <- Ok(handler.${operation.name}(request).toReactivePublisher.toStream.asJsonEither)
+              } yield responses"""
+
+          case (_: MonixObservableTpe, _: MonixObservableTpe) =>
+            q"""val requests = msg.asStream[${operation.request.safeInner}]
+              Ok(handler.${operation.name}(Observable.fromReactivePublisher(requests.toUnicastPublisher)).toReactivePublisher.toStream.asJsonEither)"""
+
+          case (_: EmptyTpe, _) =>
+            q"""Ok(handler.${operation.name}(_root_.higherkindness.mu.rpc.protocol.Empty).map(_.asJson))"""
+
+          case _ =>
+            q"""for {
+              request  <- msg.as[${operation.request.safeInner}]
+              response <- Ok(handler.${operation.name}(request).map(_.asJson)).adaptErrors
+            } yield response"""
+        }
+
+        val getPattern  = pq"GET -> Root / ${operation.name.toString}"
+        val postPattern = pq"msg @ POST -> Root / ${operation.name.toString}"
+
+        def toRouteTree: Tree = request match {
+          case _: EmptyTpe => cq"$getPattern => $routeTypology"
+          case _           => cq"$postPattern => $routeTypology"
+        }
+
+      }
+
+      val operations: List[HttpOperation] = for {
+        d      <- rpcDefs.collect { case x if findAnnotation(x.mods, "http").isDefined => x }
+        args   <- findAnnotation(d.mods, "http").collect({ case Apply(_, args) => args }).toList
+        params <- d.vparamss
+        _ = require(params.length == 1, s"RPC call ${d.name} has more than one request parameter")
+        p <- params.headOption.toList
+      } yield HttpOperation(Operation(d.name, TypeTypology(p.tpt), TypeTypology(d.tpt)))
+
+      val httpRequests = operations.map(_.toRequestTree)
+
+      val HttpClient      = TypeName("HttpClient")
+      val httpClientClass = q"""
+        class $HttpClient[$F_](uri: Uri)(implicit F: _root_.cats.effect.ConcurrentEffect[$F], ec: scala.concurrent.ExecutionContext) {
+          ..$httpRequests
+      }"""
+
+      val httpClient = q"""
+        def httpClient[$F_](uri: Uri)
+          (implicit F: _root_.cats.effect.ConcurrentEffect[$F], ec: scala.concurrent.ExecutionContext): $HttpClient[$F] = {
+          new $HttpClient[$F](uri / ${serviceDef.name.toString})
+      }"""
+
+      val httpImports: List[Tree] = List(
+        q"import _root_.higherkindness.mu.http.implicits._",
+        q"import _root_.fs2.interop.reactivestreams._",
+        q"import _root_.cats.syntax.flatMap._",
+        q"import _root_.cats.syntax.functor._",
+        q"import _root_.org.http4s._",
+        q"import _root_.org.http4s.circe._",
+        q"import _root_.io.circe._",
+        q"import _root_.io.circe.generic.auto._",
```

Comment:
IMO we should ask for the encoders/decoders implicitly and if the users want auto derivation they just need to import this.

Reply:
That makes sense. I'm on it right now but isn't being trivial, since the class now has a variable number of params (depending on the requests/responses).

Reply:
[At this commit](https://github.com/higherkindness/mu/pull/552/commits/9b0a8f3c29fe1b00b30c08db8cb695f5bbdaf252#diff-512de2cbb83d64082917cbe6d941be08L559), the implicit constraints for all the encoders and decoders have been propagated down the road, so the app who wants to run the server needs to provide implicit decoders for requests and encoders for responses, in the same way, the clients that want to perform a call need to provide implicit encoders for request and decoders for responses.